contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2070,3 +2079,242 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
 	rewrite_object_refs(stmt);
 	return res;
 }
+
+static void
+ValidateLinkedServerDataSource(char* data_src)
+{
+	/* 
+	 * Only treat fully qualified DNS names (endpoints) or IP address 
+	 * as valid data sources.
+	 * 
+	 * If data source is provided in the form of servername\\instancename, we
+	 * throw an error to suggest use of fully qualified domain name or the IP address
+	 * instead.
+	 */
+	if (strchr(data_src, '\\'))
+		ereport(ERROR,
+			(errcode(ERRCODE_FDW_ERROR),
+				errmsg("Only fully qualified domain name or IP address are allowed as data source")));
+}
+
+Datum
+sp_addlinkedserver_internal(PG_FUNCTION_ARGS)
+{
+	char *linked_server = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(0));
+	char *srv_product = PG_ARGISNULL(1) ? "" : text_to_cstring(PG_GETARG_TEXT_P(1));
+	char *provider = PG_ARGISNULL(2) ? "" : text_to_cstring(PG_GETARG_TEXT_P(2));
+	char *data_src = PG_ARGISNULL(3) ? "" : text_to_cstring(PG_GETARG_TEXT_P(3));
+	char *provstr = PG_ARGISNULL(5) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(5));
+	char *catalog = PG_ARGISNULL(6) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(6));
+
+	CreateForeignServerStmt *stmt = makeNode(CreateForeignServerStmt);
+	List *options = NIL;
+
+	bool provider_warning = false, provstr_warning = false;
+	
+	if (strlen(srv_product) == 10 && (strncmp(srv_product, "SQL Server", 10) == 0))
```

Comment:
Why the string comparison alone is not sufficient? Also, do you need to consider case-insensitivity?